qiskit/scheduler/methods/basic.py

```diff
@@ -140,31 +142,47 @@ def translate_gates_to_pulse_defs(circuit: QuantumCircuit,
     circ_pulse_defs = []
 
     cmd_def = schedule_config.cmd_def
-    measured_qubits = set()  # Collect qubits that would like to be measured
+    qubit_mem_slots = {}  # Map measured qubit index to classical bit index
 
     def get_measure_schedule() -> CircuitPulseDef:
         """Create a schedule to measure the qubits queued for measuring."""
         measures = set()
         all_qubits = set()
         sched = Schedule()
-        for q in measured_qubits:
-            measures.add(tuple(schedule_config.meas_map[q]))
+        for qubit in qubit_mem_slots:
+            measures.add(tuple(schedule_config.meas_map[qubit]))
         for qubits in measures:
             all_qubits.update(qubits)
-            # TODO (Issue #2704): Respect MemorySlots from the input circuit
-            sched |= cmd_def.get('measure', qubits)
-        measured_qubits.clear()
+            unused_mem_slots = set(qubits) - set(qubit_mem_slots.values())
+            default_sched = cmd_def.get('measure', qubits)
+            for time, inst in default_sched.instructions:
+                if isinstance(inst, AcquireInstruction):
+                    mem_slots = []
+                    for channel in inst.acquires:
+                        if channel.index in qubit_mem_slots.keys():
+                            mem_slots.append(MemorySlot(qubit_mem_slots[channel.index]))
+                        else:
+                            mem_slots.append(MemorySlot(unused_mem_slots.pop()))
+                    new_acquire = AcquireInstruction(command=inst.command,
+                                                     acquires=inst.acquires,
+                                                     mem_slots=mem_slots)
+                    sched |= new_acquire << time
+                # Measurement pulses should only be added if its qubit was measured by the user
+                elif inst.channels[0].index in qubit_mem_slots.keys():
+                    sched |= inst << time
+        qubit_mem_slots.clear()
         return CircuitPulseDef(schedule=sched, qubits=list(all_qubits))
 
-    for inst, qubits, _ in circuit.data:
+    for inst, qubits, clbits in circuit.data:
         inst_qubits = [qubit.index for qubit in qubits]  # We want only the indices of the qubits
-        if any(q in measured_qubits for q in inst_qubits):
+        if any(q in qubit_mem_slots for q in inst_qubits):
             # If we are operating on a qubit that was scheduled to be measured, process that first
             circ_pulse_defs.append(get_measure_schedule())
         if isinstance(inst, Barrier):
             circ_pulse_defs.append(CircuitPulseDef(schedule=inst, qubits=inst_qubits))
         elif isinstance(inst, Measure):
-            measured_qubits.update(inst_qubits)
+            assert len(inst_qubits) == 1 and len(clbits) == 1
```

Comment:
Instead of `assert` raising would be better. See https://github.com/Qiskit/qiskit-terra/issues/225

Reply:
Also, should it be guaranteed by construction of the instruction?

Reply:
Yes, it should be guaranteed, that was why I suggested an assert. I would usually use an assert in code where it would only be hit by an error in code, rather than a user error. For instance, if QASM ever decided to support something like `qc.measure(qubits=[0, 1], clbits=[0, 1])`, then we might want to keep this assertion statement. It's just an explicit notice of the function's assumptions. I would support raising a PulseError if it were possible for a user to build a circuit that could hit the error right now

Reply:
Ah, I see the issue there, and I am all for consistency.

Reply:
In that case..... I don't know. It seems overkill to raise an error for something we will never see, but again, if circuits change, we would end up failing silently until someone notices that their scheduled circuits aren't returning data for the right qubits

Reply:
The second case seems unlikely and easy to fix, but it's still a possibility. Is it weird to have an error in the code base solely for the purpose of raising on a future API change in another part of the code? Feels weird.

Ok, sorry for the stream of consciousness, maybe you can decide on this one.

Reply:
I've pushed [commit](https://github.com/Qiskit/qiskit-terra/pull/3415/commits/f55094bd45eab3025f93e4ba99a27f59402979b0) to remove `assert` to `if (condition): raise QiskitError`.